File: 2022/cpp/day_4.cpp
#include "headers/stringandvectorutils.h"
#include <fstream>
#include <iostream>
#include <vector>

void part_1();
void part_2();
std::vector<std::string> parse_input();

void part_1() {
  std::vector<std::string> day_data = parse_input();
  int sumr = 0;

  for (size_t i = 0; i < day_data.size(); i++) {
    std::string item = day_data[i];
    std::vector<std::string> vec_list = split_string_to_list(item, ',');

    std::vector<std::string> shift_1_s = split_string_to_list(vec_list[0], '-');
    std::vector<std::string> shift_2_s = split_string_to_list(vec_list[1], '-');

    std::vector<int> shift_1 = vector_string_to_int(shift_1_s);
    std::vector<int> shift_2 = vector_string_to_int(shift_2_s);

    if ((shift_1[0] <= shift_2[0] && shift_1[1] >= shift_2[1]) ||
        (shift_2[0] <= shift_1[0] && shift_2[1] >= shift_1[1])) {
      sumr += 1;
    }
  }

  std::cout << "The answer to part 1 is: " << sumr << std::endl;
}

void part_2() {
  std::vector<std::string> day_data = parse_input();
  int sumr = 0;

  for (size_t i = 0; i < day_data.size(); i++) {
    std::string item = day_data[i];
    std::vector<std::string> vec_list = split_string_to_list(item, ',');

    std::vector<std::string> shift_1_s = split_string_to_list(vec_list[0], '-');
    std::vector<std::string> shift_2_s = split_string_to_list(vec_list[1], '-');

    std::vector<int> shift_1 = vector_string_to_int(shift_1_s);
    std::vector<int> shift_2 = vector_string_to_int(shift_2_s);

    if ((shift_1[0] <= shift_2[0] && shift_1[1] >= shift_2[1]) ||
        (shift_2[0] <= shift_1[0] && shift_2[1] >= shift_1[1])) {
      sumr += 1;
    } else if (shift_1[1] >= shift_2[0] && shift_1[1] <= shift_2[1]) {
      sumr += 1;
    } else if (shift_1[0] >= shift_2[0] && shift_1[0] <= shift_2[1]) {
      sumr += 1;
    }
  }

  std::cout << "The answer to part 2 is: " << sumr << std::endl;
}

int main() {
  part_1();
  part_2();
  return 0;
}

std::vector<std::string> parse_input() {
  std::vector<std::string> nums;
  std::ifstream f("Inputs/day_4_input.txt");

  std::string s;

  while (getline(f, s)) {
    nums.push_back(s);
  }

  f.close();

  return nums;
}

File: 2022/Python/day_4.py
def parse_input():
    splitter = []
    with open("Inputs/day_4_input.txt") as input_file:
        for line in input_file:
            ltext = line.replace("\n", "")
            splitter.append(ltext)
    return splitter


def part_1():
    data = parse_input()
    sumr = 0
    for i in data:
        split_e = i.split(",")
        e1 = split_e[0].split("-")
        e2 = split_e[1].split("-")

        e1 = [int(x) for x in e1]
        e2 = [int(x) for x in e2]

        if (e1[0] <= e2[0] and e1[1] >= e2[1]) or (e2[0] <= e1[0] and e2[1] >= e1[1]):
            sumr += 1

    print("The total entirely overlapping assignments for Part 1 is: " + str(sumr))


def part_2():
    data = parse_input()
    sumr = 0
    for i in data:
        split_e = i.split(",")
        e1 = split_e[0].split("-")
        e2 = split_e[1].split("-")

        e1 = [int(x) for x in e1]
        e2 = [int(x) for x in e2]

        if (e1[0] <= e2[0] and e1[1] >= e2[1]) or (e2[0] <= e1[0] and e2[1] >= e1[1]):
            sumr += 1
        elif e1[1] >= e2[0] and e1[1] <= e2[1]:
            sumr += 1
        elif e1[0] >= e2[0] and e1[0] <= e2[1]:
            sumr += 1

    print("The total overlapping assignments for Part 2 is: " + str(sumr))


if __name__ == "__main__":
    part_1()
    part_2()
